src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -624,6 +624,39 @@ int D4NFilterBucket::list(const DoutPrefixProvider* dpp, ListParams& params, int
   return 0;
 }
 
+int D4NFilterBucket::remove(const DoutPrefixProvider* dpp,
+			    bool delete_children,
+			    optional_yield y)
+{
+  ListParams params;
+  ListResults results;
+  cache_request = true; // though not a cache request, this helps limit the scope of the list call to what is in the cache
+  int ret = list(dpp, params, 1000, results, y);
+  if (ret < 0) {
+    return ret;
+  }
+
+  for (const auto& obj : results.objs) { 
+    ldpp_dout(dpp, 20) << "D4NFilterBucket::" << __func__ << "(): deleting object " << obj.key << dendl;
+    std::unique_ptr<rgw::sal::Object> c_obj = this->get_object(obj.key);
+    ACLOwner owner{this->get_owner()}; // TODO: Is this ownership derivation correct in this case?
```

Comment:
comment in code

Reply:
del_op->params.obj_owner = owner;  appears to be ok, but del_op->params.bucket_owner = this->get_owner(); should be the bucket owner (use get_owner() method of FilterBucket.

Reply:
The cache API PR is a separate PR from this but since it uses the `cache_request` bool, I have the commits in this PR as well. I am working on getting it merged and will rebase this PR once that occurs so it is only related to bucket removal changes.

Reply:
I meant pipelining should be used for entries in the cache. Lets say there are 1000 entries in the cache. If you issue a delet_obj call for each entry, it will be very inefficient. So you can use pipelining to check for existence of entries in the cache (you could probably ignore this step, as you are already listing its entries) 2. write a new method that uses pipelining to delete corresponding redis entries. (You can check get and set methods in directory.cc that use pipelining)

Reply:
> Do you mean pass the delete_obj call for that object to the backend store? Why is this necessary for objects that are not in the cache when we are calling next->remove at the end of the method regardless? The backend store would handle deleting its objects during this call, so in cases where the object doesn't exist, shouldn't the logic simply continue? Please correct my understanding if it is wrong

Yes you do not need to call delete_obj for every backend entry, bucket->remove will take care.